File: ffplaylists/ffplaylists/PlaylistBuilder.m
//
//  PlaylistBuilder.m
//  ffplaylists
//
//  Created by Peter Kollath on 13/11/14.
//  Copyright (c) 2014 Peter Kollath. All rights reserved.
//

#import "PlaylistBuilder.h"
#import "FlatFileUtils.h"

@implementation PlaylistBuilder

-(id)init
{
    self = [super init];
    if (self)
    {
        for(NSInteger i = 0; i < kMaxLevel; i++)
        {
            levelRecIds[i] = -1;
            levelRecs[i] = nil;
            self.gid = 1;
            self.printedRecs = [NSMutableDictionary new];

        }
    }
    return self;
}


-(BOOL)validate
{
    return (self.inputFile.length > 0 && self.outputDir.length > 0 && self.levelFile.length > 0);
}

-(void)loadLevels
{
    self.levels = [NSMutableDictionary new];
    NSString * str = [NSString stringWithContentsOfFile:self.levelFile
                                               encoding:NSUTF8StringEncoding
                                                  error:NULL];
    NSArray * lines = [str componentsSeparatedByCharactersInSet:[NSCharacterSet newlineCharacterSet]];
    for (NSString * line in lines)
    {
        NSArray * parts = [line componentsSeparatedByString:@"\t"];
        if (parts.count == 3)
        {
            [[self levels] setObject:[NSNumber numberWithInt:[(NSString *)parts[0] intValue]]
                              forKey:parts[2]];
        }
    }
}


-(void)process
{
    [self loadLevels];
    
    int count = 0;
    int len = 0;
    FILE * i = fopen([self.inputFile UTF8String], "rt");
    
    self->outputFile = fopen([[self.outputDir stringByAppendingPathComponent:@"playlists.txt"] UTF8String], "wt");
    self->outputFile2 = fopen([[self.outputDir stringByAppendingPathComponent:@"playlists_detail.txt"] UTF8String], "wt");
    
    char *c = malloc(200000);
    while((c = fgets(c, 200000, i)) != NULL)
    {
        len = (int)strlen(c);
        if (len > 0 && c[len-1] == '\n')
            c[len-1] = '\0';
        NSString * line = [NSString stringWithUTF8String:c];
        NSArray * part = [line componentsSeparatedByString:@"\t"];
        
        [self processText:part[1] recordId:[(NSString *)part[0] intValue] styleName:part[3] levelName:part[2]];
        count ++;
        
        if (count % 1000 == 0) {
            NSString * str = [NSString stringWithFormat:@"processed %d lines", count];
            printf("%s", str.UTF8String);
            for(int i = 0; i < str.length; i++)
            {
                printf("\b");
            }
            fflush(stdout);
        }
    }
    
    fclose(i);
    fclose(self->outputFile);
    fclose(self->outputFile2);
}

-(NSUInteger)getLevelIndex:(NSString *)levelName
{
    if (levelName == nil || [levelName length] == 0)
        return NSNotFound;
    
    NSNumber * index = [self.levels objectForKey:levelName];
    if (index != nil) {
        return [index unsignedIntegerValue];
    }
    
    return NSNotFound;
}

-(void)processText:(NSString *)text
          recordId:(int)recId
         styleName:(NSString *)style
         levelName:(NSString *)levelName
{
    NSUInteger level = [self getLevelIndex:levelName];
    
    if (level != NSNotFound && level <= self.maxLevel)
    {
        for(NSInteger i = level; i < kMaxLevel; i++)
        {
            levelRecIds[i] = -1;
            levelRecs[i] = nil;
        }
    }
    
    if ([style isEqualToString:@"PA_Audio_Bg"])
    {
        NSRange r1, r2;
        
        r1 = [text rangeOfString:@"<AUDIO:\""];
        if (r1.location == NSNotFound)
        {
            r1 = [text rangeOfString:@"<DL:Data,\""];
        }
        
        if (r1.location != NSNotFound)
        {
            r2 = [text rangeOfString:@"\"" options:0 range:NSMakeRange(r1.location + r1.length, text.length - r1.location - r1.length)];
        }
        
        NSString * object = nil;
        
        if (r1.location != NSNotFound && r2.location != NSNotFound)
        {
            object = [text substringWithRange:NSMakeRange(r1.location + r1.length, r2.location - r1.location - r1.length)];
        }
        
        if (object)
        {
            NSInteger prevLevId = -1;
            for(NSInteger i = 0; i < kMaxLevel; i++)
            {
                if (levelRecIds[i] == -1)
                    continue;
                NSString * key = [NSString stringWithFormat:@"%ld", levelRecIds[i]];
                if (![self.printedRecs valueForKey:key])
                {
                    NSString * title = [FlatFileUtils removeTags:levelRecs[i]];
                    // PLAYLT  levelRecs[i]   prevLevId     title
                    fprintf(self->outputFile, "%ld\t%ld\t%s\n", levelRecIds[i], prevLevId, [title UTF8String]);
                    [self.printedRecs setValue:@"1" forKey:key];
                }
                
                prevLevId = levelRecIds[i];
            }
            
            fprintf(self->outputFile2, "%ld\t%ld\t%s\n", prevLevId, self.gid, [object UTF8String]);
            //OBJECT   self.gid   prevLevId    object
            self.gid++;
        }
    }
    else
    {
        if (level != NSNotFound && level <= self.maxLevel)
        {
            self->levelRecs[level] = text;
            self->levelRecIds[level] = self.gid;
            self.gid++;
        }
    }
}


@end

File: python/PlaylistBuilder.py
import os
import os.path

import Flat

kMaxLevel = 100

class PlaylistBuilder:
    def __init__(self):
        self.outputFile = None
        self.outputFile2 = None
        self.levelRecs = [None] * kMaxLevel
        self.levelRecIds = [-1] * kMaxLevel
        self.outputDir = None
        self.inputFile = None
        self.levelFile = None
        self.levels = None
        self.maxLevel = 6
        self.gid = 1
        self.regex1 = None
        self.regex2 = None
        self.printedRecs = {}

    def loadLevels(self):
        self.levels = {}
        with open(self.levelFile,'rt',encoding='utf-8') as rf:
            for line in rf:
                parts = line.strip().split('\t')
                if len(parts)==3:
                    self.levels[parts[2]] = int(parts[0])

    def validate(self):
        return len(self.inputFile) > 0 and len(self.outputDir) > 0 and len(self.levelFile) > 0

    def getLevelIndex(self,levelName):
        if levelName == None or len(levelName)==0: return -1
        if levelName not in self.levels: return -1
        return self.levels[levelName]

    def process(self):
        self.loadLevels()

        count = 0

        self.outputFile = open(os.path.join(self.outputDir, "playlists.txt"), "wt", encoding='utf-8')
        self.outputFile2 = open(os.path.join(self.outputDir, "playlists_details.txt"), "wt", encoding='utf-8')

        with open(self.inputFile,'rt',encoding='utf-8') as input_file:
            for line in input_file:
                part = line.strip('\n').split('\t')

                self.processText(part[1], int(part[0]), part[3], part[2])
                count += 1

                if count % 2000 == 0:
                    print(f'\rProcessed {count} lines.',end='')
        print()

        self.outputFile.close()
        self.outputFile2.close()


    def processText(self,text,recId,style,levelName):
        level = self.getLevelIndex(levelName)

        if level != -1 and level <= self.maxLevel:
            for i in range(level,kMaxLevel):
                self.levelRecs[i] = None
                self.levelRecIds[i] = -1

        if style=="PA_Audio_Bg":
            #print(recId, text)
            r1 = text.find("<AUDIO:\"")
            r1l = 8
            if r1<0:
                r1 = text.find("<DL:Data,\"")
                r1l = 10

            if r1>=0:
                r2 = text.find('"', r1 + r1l)

            if r1>=0 and r2 > r1:
                object = text[r1 + r1l:r2]
                prevLevId = -1
                for i in range(kMaxLevel):
                    if self.levelRecIds[i] == -1: continue
                    key = self.levelRecIds[i]
                    if key not in self.printedRecs:
                        # PLAYLT  levelRecs[i]   prevLevId     title
                        print("{}\t{}\t{}".format(self.levelRecIds[i], prevLevId, Flat.removeTags(self.levelRecs[i])), file=self.outputFile)
                        self.printedRecs[key] = True
                    prevLevId = self.levelRecIds[i]

                print("{}\t{}\t{}".format(prevLevId, self.gid, object), file=self.outputFile2)
                #OBJECT   self.gid   prevLevId    object
                self.gid+=1
        else:
            if level != -1 and level <= self.maxLevel:
                self.levelRecs[level] = text
                self.levelRecIds[level] = self.gid
                self.gid+=1
